pkg/controller/daemon/update_test.go

```diff
@@ -117,7 +121,190 @@ func TestDaemonSetUpdatesPodsWithMaxSurge(t *testing.T) {
+	}
+	setDaemonSetToleration(ds, tolerations)
+	manager, podControl, _, err := newTestController(ctx, ds)
+	if err != nil {
+		t.Fatalf("error creating DaemonSets controller: %v", err)
+	}
+	err = manager.dsStore.Add(ds)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Add five nodes and taint to one node
+	addNodes(manager.nodeStore, 0, 5, nil)
+	taints := []v1.Taint{
+		{Key: "node-role.kubernetes.io/control-plane", Effect: v1.TaintEffectNoSchedule},
+	}
+	node := newNode("node-0", nil)
+	setNodeTaint(node, taints)
+	err = manager.nodeStore.Update(node)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create DaemonSet with toleration
+	expectSyncDaemonSets(t, manager, ds, podControl, 5, 0, 0)
+	markPodsReady(podControl.podStore)
+
+	// RollingUpdate DaemonSet without toleration
+	ds.Spec.Template.Spec.Containers[0].Image = "foo2/bar2"
+	ds.Spec.Template.Spec.Tolerations = nil
+	err = manager.dsStore.Update(ds)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, maxSurge, 1, 0)
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
+	markPodsReady(podControl.podStore)
+
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, maxSurge, maxSurge, 0)
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
+	markPodsReady(podControl.podStore)
+
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, maxSurge, maxSurge, 0)
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
+	markPodsReady(podControl.podStore)
+
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, maxSurge, maxSurge, 0)
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
+	markPodsReady(podControl.podStore)
+
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, maxSurge, 0)
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
+}
+
+func TestDaemonSetUpdatesNodeIntersectionWithMaxSurge(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+	var tolerations []v1.Toleration
+	var taints []v1.Taint
+	var node *v1.Node
+
+	ds := newDaemonSet("foo")
+	maxSurge := 1
+	ds.Spec.UpdateStrategy = newUpdateSurge(intstr.FromInt(maxSurge))
+	tolerations = []v1.Toleration{
+		{Key: "zone-y", Operator: v1.TolerationOpExists},
+	}
+	setDaemonSetToleration(ds, tolerations)
+	manager, podControl, _, err := newTestController(ctx, ds)
+	if err != nil {
+		t.Fatalf("error creating DaemonSets controller: %v", err)
+	}
+	err = manager.dsStore.Add(ds)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Add four nodes with zone-y or zone-z taint
+	for i := 0; i < 4; i++ {
```

Comment:
also as I suggested in https://github.com/kubernetes/kubernetes/pull/119317#discussion_r1272431930, might be worth to have more nodes and have some intersection between them

Reply:
I'm not sure I fully understand what you're saying, but your suggesting test scenario in integration test would be something like that?

[step0] Prepare 6 Nodes with taint as below:
```
node-0, node-1
  Key: "zone-y", Effect: NoSchedule
node-2, node-3
  Key: "zone-z", Effect: NoSchedule
node-4, node-5  <- intersection between zone-y and zone-z
  Key: "common", Effect: NoSchedule
```

[step1] Create DaemonSet with these tolerations and expect 4 Pods to run.
```  
Key: "zone-y", Operator: Exists
Key: "common", Operator: Exists
```

[step2] Update DaemonSet with these tolerations and expect 4 Pods to run.
```
Key: "zone-z", Operator: Exists
Key: "common", Operator: Exists
```
	
[step3] Update DaemonSet with these tolerations and expect 6 Pods to run.
```  
Key: "zone-y", Operator: Exists
Key: "zone-z", Operator: Exists
Key: "common", Operator: Exists
```

[step4] Update DaemonSet with no toleration and expect 0 Pod to run.

Reply:
yup, that should work for our use case - to see some of the pods to go without rescheduling

Reply:
OK, I've added the new integration test case.

Reply:
@mochizuki875 thanks for moving it to the integration tests